### DISCUSSION TITLE
Create Taxonomy Programmatically in Migration

### DISCUSSION THREAD

**Discussion by fuzl-llc** [OrchardCMS/OrchardCore#12226] [Q&A]
I'd like to create a taxonomy during a migration. I can create the taxonomy's "term content type" as so:

```
_contentDefinitionManager.AlterTypeDefinition("Subject", type => type
    .WithPart(nameof(TitlePart))
);
```

Then, I try to create my taxonomy:

```
var subjectsTaxonomyContentItem = await _contentManager.NewAsync("Taxonomy");
subjectsTaxonomyContentItem.DisplayText = "Subjects";
var subjectsTaxonomyPart = subjectsTaxonomyContentItem.As<TaxonomyPart>();
subjectsTaxonomyPart.TermContentType = "Subject"; // ERROR: subjectsTaxonomyPart == null
await _contentManager.CreateAsync(subjectsTaxonomyContentItem, VersionOptions.Published);
```

So, the taxonomy part being null seems to be the issue. I've seen #3721 which the above is based on but that code is 3 years old. It  also has `taxonomy = taxonomy.Apply(taxonomyPart)` but I'm not sure if that is needed and it would also fail due to taxonomy part being null. With the line commented out, it does create the taxonomy but the term content type of course isn't set. Neither is the name/technical name and I see no way to set them. I can re-publish from the UI setting all those but that sort of defeats the point of the migration.

Is there a way to create a taxonomy programmatically from a migration?

**Comment by tvsbrent** [comment]
The way I addressed this was to add the taxonomy part to the content type. Example:

```csharp
  _contentDefinitionManager.AlterTypeDefinition(nameof(ContentTypes.Subject), type => type
      .Listable()
      .WithPart(nameof(TitlePart))
      .WithPart(nameof(TaxonomyPart))
  );
```

Then the casting as `TaxonomyPart` worked.

**EDIT:** The above does not work, apologies for misleading. Creating your own content type with a `TaxonomyPart` does work, and you can do the casting as a `TaxonomyPart` in the migration. However, you do not seem able to assign a content item created in this way to a taxonomy field later, at least in the UI. I haven't tried it programmatically yet.

**Comment by fuzl-llc** [comment]
Thanks anyway @tvsbrent. I added the `.WithPart(nameof(TaxonomyPart))` but I still get `subjectsTaxonomyPart == null`

I guess the TaxonomyPart should be on `subjectsTaxonomyContentItem`, not the "Subject" type anyway, since `subjectsTaxonomyContentItem` is the actual taxonomy and "Subject" is the term content type for that taxonomy type.
